Enhance API-Testing-Guidelines.md (#203)
camaraproject/Commonalities: documentation/API-Testing-Guidelines.md

# API Testing Guidelines

## Table of Contents
- [Introduction](#introduction)
- [Test Cases contributions for APIs](#contribution)
- [Location of feature file](#location)
- [Test plan design guidelines](#guidelines)
  - [Feature structure](#feature-structure)
  - [Environment variables](#environment-variables)
  - [Background section](#background)
  - [Scenario structure](#scenario-structure)
  - [Request setup](#request-setup)
  - [Request sending](#request-sending)
  - [Response validation](#response-validation)
- [Test plan coverage](#coverage)
  - [Independent operations](#independent-operations)
  - [CRUD operations](#crud-operations)
  - [Notifications](#notifications)
  - [Error scenarios](#error-scenarios)
- [References](#references)

## Introduction <a name="introduction"></a>
This document captures guidelines for the API testing in CAMARA project. These guidelines are applicable to every API to be worked out under the CAMARA initiative.

## Test Cases contribution for APIs <a name="contribution"></a>

> Based on the decision taken in the Commonalities working group and as documented in [API-Readiness-Checklist](https://github.com/camaraproject/Commonalities/blob/main/documentation/API-Readiness-Checklist.md), at least one Gherkin feature file will be added to the main subproject repo and this will fulfill the minimum criteria of readiness with respect to API test cases and documentation. If subprojects also intend to add test implementations, an aligned single implementation that is agreed among all provider implementors could be added to the main subproject repo. If no alignment is possible, each provider implementer will add the test implementation to their own repos.

Reviewer: The updated API Readiness Checklist is proposed in Release Msanagement: https://github.com/camaraproject/ReleaseManagement/pull/29

It impacts this section and possibly "Basic API test cases & documentation" and "Enhanced API test cases & documentation"  should be defined here.

Author: It's difficult to set the difference between Basic and Enhanced. To me basic tests cases should cover at least that the implementation interface is compliant with the API spec, in terms of syntax and support for all mandatory aspects. Enhanced test plan may validate unusual error scenarios and corner cases. But this is very subtle.

## Location of feature files <a name="location"></a>

The feature files will reside under 

`{Subproject_Repository}/code/Test_definitions/file_name.feature`

E.g.: 

`https://github.com/camaraproject/DeviceLocation/blob/main/code/Test_definitions/location-verification.feature`

Filename:

- APIs with a single feature file will use the api-name as file name: E.g. `location-verification.feature`
- For APIs with several operations that split the test plan in one feature file per operation, recommendation is to use the `operationId` as file name: `createSession.feature`.

> - For APIs with several feature files, not split by operation, may use the api-name along some other description, e.g.: `qod-notifications.feature`

Author: Split is right I think :)

* the headers:
    | param_name | param_value |
    | ---------- | ----------- |
    | xxx        | yyy         |
* the query parameter "{name}" is set as "{value}"
* the query parameters:
    | param_name | param_value |
    | ---------- | ----------- |
    | xxx        | yyy         |
* the request body property "{json_path}" is set as "{value}"
* the request body:
    ```
    {
      "xxx": "yyy"
    }
    ```
```

To refer to properties in request bodies and response, [JSON path](https://www.ietf.org/archive/id/draft-goessner-dispatch-jsonpath-00.html) notation is recommended.

Alternatively, instead of specifying exact values for specific parameters or properties, testing plans can describe the value with a human-friendly sentence, to let the API tester fill the request with some value that complies with the requirement, e.g.:

```
* the path parameter "id" is that of an existing item
* the query parameter "startDate" is in the past
* the request body property "$.device" is a valid device which location is known
```

Testing plans should identify the list of input values that have to be provided by the implementation.

### Request sending <a name="request-sending"></a>

> For scenarios testing individual operations, the request will be send by means of a single `When` step, indicating the `operationId` specified in the spec, as the `operationId` uniquely identifies the HTTP method and resource path. E.g.:

Reviewer 2: ```suggestion
For scenarios testing individual operations, the request will be sent by means of a single `When` step, indicating the `operationId` specified in the spec, as the `operationId` uniquely identifies the HTTP method and resource path. E.g.:
```

# API Testing Guidelines

## Table of Contents
- [Introduction](#introduction)
- [Test plan coverage](#coverage)
  - [Independent operations](#independent-operations)
  - [CRUD operations](#crud-operations)
  - [Notifications](#notifications)
  - [Error scenarios](#error-scenarios)
- [Test plan design guidelines](#guidelines)
  - [Location of feature file](#location)
  - [Feature structure](#feature-structure)
  - [Environment variables](#environment-variables)
  - [Background section](#background)
  - [Scenario structure](#scenario-structure)
  - [Request setup](#request-setup)
  - [Request sending](#request-sending)
  - [Response validation](#response-validation)
- [References](#references)

## Introduction <a name="introduction"></a>
This document captures guidelines for the API testing in CAMARA project. These guidelines are applicable to every API to be worked out under the CAMARA initiative.

## Test plan coverage <a name="coverage"></a>

> Based on the decision taken in the Release Management and Commonalities working groups, and as documented in [API-Readiness-Checklist](https://github.com/camaraproject/Commonalities/blob/main/documentation/API-Readiness-Checklist.md), at least one Gherkin feature file will be added to the main subproject repo and this will fulfill the minimum criteria of readiness with respect to API test cases and documentation:

Reviewer: I propose to change the link to Release Management wiki.
Current API Readiness Checklist will be deprecated: https://github.com/camaraproject/Commonalities/issues/202
```suggestion
Based on the decision taken in the Release Management and Commonalities working groups, and as documented in [API Release Process](https://wiki.camaraproject.org/display/CAM/API+Release+Process)](https://wiki.camaraproject.org/display/CAM/API+Release+Process#APIReleaseProcess-APIreadinesschecklist), at least one Gherkin feature file will be added to the main subproject repo and this will fulfill the minimum criteria of readiness with respect to API test cases and documentation:
```

> Based on the decision taken in the Release Management and Commonalities working groups, and as documented in [API-Readiness-Checklist](https://github.com/camaraproject/Commonalities/blob/main/documentation/API-Readiness-Checklist.md), at least one Gherkin feature file will be added to the main subproject repo and this will fulfill the minimum criteria of readiness with respect to API test cases and documentation:

Author: We can clarify that explicitly that, but the "API Readiness Checklist" applies to each API individually, so yes, each API which is part of a meta-release, regardless whether it is sharing a subproject with other APIs, must have its own test plan.

* Any public release must include a full test plan, covering both sunny and rainy day scenarios, to be considered stable. This may include unusual or not explicitly specified error scenarios, semantic validations and corner cases.

If subprojects also intend to add test implementations, an aligned single implementation that is agreed among all provider implementors could be added to the main subproject repo. If no alignment is possible, each provider implementer will add the test implementation to their own repos.

* Test plans are not expected to certify or measure the quality of the implementation in terms of accuracy or veracity of the response, but to validate that the interface is compliant with the API specification and that the API is fully implemented.

* When the API allows options or alternative approaches, the test plan should cover those situations and define the expected behaviour for inputs or options not supported.
  - Implementations will need to inform in advance to the API tester about which subset of options are supported. 

### Independent operations <a name="independent-operations"></a>

> Operations that execute an action or retrieve information from the operator, without modifying a resource, will be tested in its own dedicated feature file. 

Reviewer 3: Does this mean we will need two different feature files for QoD Operations ? If yes then what would be significance of it ?

Author: QoD is mostly a CRUD API, encapsulating sessions resources, so it lies on the other category. Anyway, it is up to QoD subproject, and the contributors of its test plans, to decide on how many feature files are convenient. Splitting a test plan in several features is a test plan design decision, taking into account the complexity of the test plan. The more complex a test plan is, the most likely is that it will be split into several feature files. As in any software project.

These guidelines are recommendations for the design, so we may make the text less normative: "will be tested in its own dedicated feature file" -> "will be **usually** tested in its own dedicated feature file"

Reviewer 3: @jlurien  OK with us kindly accommodate this in PR file as well

Author: Added

> Operations that execute an action or retrieve information from the operator, without modifying a resource, will be tested in its own dedicated feature file. 

Reviewer: ```suggestion
Operations that execute an action or retrieve information from the operator, without modifying a resource, will be usually tested in its own dedicated feature file. 
```

Test plan must validate that providing a valid input, a success response is received and that the response body complies with the JSON-schema in the API spec.

For operations receiving an input request body with optional properties, several scenarios should be included, testing different sets of valid inputs.

> ### CRUD operations <a name="crud-operations"></a>

Reviewer 3: Can we rename this to Resource Modifying Operations to make the differentiation clearer

Author: Fine to me

Reviewer 3: @jlurien  Kindly accommodate in PR file as well

> ### CRUD operations <a name="crud-operations"></a>

Reviewer: ```suggestion
### Resource modifying operations <a name="crud-operations"></a>
```

> For path resources with CRUD (Create/Read/Update/Delete) operations, some scenario should validate that the new state of the resource is coherent with the response to the operation.

Reviewer 3: * response -> request correction

Author: I can clarify this further, but it really meant the **response**. The request may ask the server for something, but it is the response which tells the client whether the server has granted the request and therefore modify the state of the resource accordingly.

Reviewer 3: @jlurien It seems new state check is missing in the detailed steps below

Author: Text rephrased

* For Update operations:
  - If operation success returns the updated object, validate that the properties of the response object are coherent with the request object
  - Validate that the updated object can be read by the GET operation for the item id, with the modified properties.

* For Delete operations:
  - Validate that trying to read the successfully deleted object, with the GET operation for the deleted item id, returns the error specified in the API spec, typically 404.

### Notifications <a name="notifications"></a>

For operation with implicit subscriptions:

* Check that when a webhook is provided, the expected events are received in the `notificationUrl`, with the right `notificationAuthToken`, for those situations specified in the API.
* If the API allows to update a previously provided webhook:
  - If the `notificationUrl` is modified, validate that events are received in the modified value.
  - If the `notificationUrl` can be nullified, validate that events are not longer received.
  - If the `notificationAuthToken` is modified, validate that the new token is used for the events.

For the explicit subscriptions model:

> * Validate that the subscribed events are received in the `notificationUrl`, with the right `notificationAuthToken`, for those situations specified in the API.

Reviewer 3: @jlurien  Consideration of new template  event-subscription-template.yaml needed in Testing-Guidelines as well

Author: I don't get this comment, sorry. Could you please elaborate further? Which template and which considerations?

Reviewer 3: @jlurien https://github.com/camaraproject/Commonalities/pull/189 This was new template introduced & consideration of this for Testing guidelines will be useful cc @akoshunyadi

Reviewer: @jlurien The template for explicit event subscription (like geofencing): https://github.com/camaraproject/Commonalities/blob/main/artifacts/camara-cloudevents/event-subscription-template.yaml

Author: It should be aligned now after @PedroDiez's suggestions. Please double check

* For subscriptions that provide `subscriptionExpireTime`, validate that the subscribed events are not longer received after the expiration time.
* For subscriptions that provide `subscriptionMaxEvents`, validate that the subscribed events are not longer received after the maximum events limit is reached..
* Validate that after a subscription is deleted, the subscribed events are not longer received.

### Error scenarios <a name="error-scenarios"></a>

* All errors explicitly documented in the API spec must be covered by one or more dedicated error scenarios.

>   - In particular operations that expect a `device` object allowing one or more identifiers to be provided, must include scenarios to test all error cases defined. 

Reviewer 3: @jlurien Can you please clarify why only device object is considered explicitly? We think generic phrase would be fine for covering error scenarios.

Author: Because recently errors for device processing have been discussed specifically, and new design guidelines also have a dedicated mention to device error management. But this do not exclude any other error scenario. Idea is to give an example of deep of tests that are expected,

Author: added reference to the device error management section of the API design guidelines

- Values with invalid format
- Mandatory properties are missing

## Test plan design guidelines <a name="guidelines"></a>

* Granularity of the feature file must be decided at the project level but it is recommended to:
    -	group in one file all scenarios testing one closely related API capability (that can cover one or several endpoints).
    -	provide several files when one CAMARA API (yaml) covers several independent functions that can be provided independently

* Third person pronoun usage in feature file is advisable as using the third person, conveys information in a more official manner.

### Location of feature files <a name="location"></a>

The feature files will reside under 

`{Subproject_Repository}/code/Test_definitions/file_name.feature`

E.g.: 

`https://github.com/camaraproject/DeviceLocation/blob/main/code/Test_definitions/location-verification.feature`

Filename:

- APIs with a single feature file will use the api-name as file name: E.g. `location-verification.feature`
- For APIs with several operations that split the test plan in one feature file per operation, recommendation is to use the `operationId` as file name: `createSession.feature`.

> - For APIs with several feature files, not split by operation, may use the api-name along some other description, e.g.: `qod-notifications.feature`

Reviewer 3: @jlurien  We think API name should be mandatory with some added description if needed .

Author: Fine to me, I could suggest having always the api-name and, when the testing plan is composed of several files, that suffix, which can be operationId or other description.

Author: Suggestion from @rartych applied

Reviewer 3: @jlurien  File still does not contain apiname-operationid/description in 2nd point  for Createsession.feature.
Kindly amend in the PR

Reviewer: I think adding api-name in the second point brings consistency (filename always starts with api-name) , but can be helpful in the future with multiple API specifications in the same repository (distinguishing tests in case the same operationId is used).

> - For APIs with several operations that split the test plan in one feature file per operation, recommendation is to use the api-name along the `operationId` as file name: `device-identifier-retrieveType.feature`.

Author: I agree. Applied

> - For APIs with several feature files, not split by operation, may use the api-name along some other description, e.g.: `qod-notifications.feature`

Reviewer: ```suggestion
- For APIs with several feature files, not split by operation, should use the api-name along some other description, e.g.: `qod-notifications.feature`
```

Author: I rephrased the sentence
